quotaservice-cli/cli.go

```diff
@@ -65,38 +72,106 @@ func main() {
-	fmt.Println(string(body))
+
+	if *output == "" {
+		fmt.Print(string(body))
+	} else {
+		logf("Writing to %v\n", *output)
+		f, err := os.Create(*output)
+		kingpin.FatalIfError(err, "Cannot write to file %v", *output)
+		f.WriteString(string(body))
+		f.Close()
+	}
 }
 
 func doAdd(gdb bool, namespace, bucket string) {
 	validate(gdb, namespace, bucket)
-	logf("Called add(%v, %v, %v)\n", gdb, namespace, bucket)
-	cfgBytes, e := ioutil.ReadAll(os.Stdin)
-	kingpin.FatalIfError(e, "Could not read config from stdin")
-	cfg := string(cfgBytes)
-	logf("Read config %v from stdin\n", cfg)
-	_ = createUrl(gdb, namespace, bucket)
-	// TODO: Make REST call to server to add config and print status
+	logf("Called add(gdb=%v, namespace=%v, bucket=%v)\n", gdb, namespace, bucket)
+	cfgBytes := readCfg(*addFile, namespace, bucket)
+	url := createUrl(gdb, namespace, bucket)
+	resp := connectToServer("POST", url, cfgBytes)
+	resp.Body.Close()
 }
 
 func doRemove(gdb bool, namespace, bucket string) {
 	validate(gdb, namespace, bucket)
-	logf("Called remove(%v, %v, %v)\n", gdb, namespace, bucket)
-	_ = createUrl(gdb, namespace, bucket)
-	// TODO: Make REST call to server to remove config and print status
+	logf("Called remove(gdb=%v, namespace=%v, bucket=%v)\n", gdb, namespace, bucket)
+	url := createUrl(gdb, namespace, bucket)
+	resp := connectToServer("DELETE", url)
+	resp.Body.Close()
 }
 
 func doUpdate(gdb bool, namespace, bucket string) {
 	validate(gdb, namespace, bucket)
-	logf("Called update(%v, %v, %v)\n", gdb, namespace, bucket)
-	_ = createUrl(gdb, namespace, bucket)
-	// TODO: Make REST call to server to update config and print status
+	logf("Called update(gdb=%v, namespace=%v, bucket=%v)\n", gdb, namespace, bucket)
+	cfgBytes := readCfg(*updateFile, namespace, bucket)
+	url := createUrl(gdb, namespace, bucket)
+	resp := connectToServer("PUT", url, cfgBytes)
+	resp.Body.Close()
+}
+
+func readCfg(f, namespace, bucket string) []byte {
+	var cfgBytes []byte
+	var e error
+
+	if f == "" {
+		f = "STDIN"
+		cfgBytes, e = ioutil.ReadAll(os.Stdin)
+	} else {
+		cfgBytes, e = ioutil.ReadFile(f)
+	}
+
+	kingpin.FatalIfError(e, "Could not read config from %v", f)
+	logf("Read config %v from %v\n", string(cfgBytes), f)
+	validateJSON(cfgBytes, namespace, bucket)
+	return cfgBytes
+}
+
+func validateJSON(j []byte, namespace, bucket string) {
+	var js map[string]interface{}
+	if json.Unmarshal(j, &js) != nil {
+		kingpin.Fatalf("Config read isn't valid JSON!\n")
+	}
+
+	if namespace == "" {
+		// Global default bucket
+		checkField("name", config.DefaultBucketName, js, func(val string) {
+			kingpin.Fatalf("Global default bucket cannot have name '%v'", val)
+		})
+		checkField("namespace", config.GlobalNamespace, js, func(val string) {
+			kingpin.Fatalf("Global default bucket cannot have namespace '%v'", val)
+		})
+	} else if bucket == "" {
+		// We're just updating a namespace.
+		checkField("name", namespace, js, func(val string) {
```

Comment:
The JSON config of a namespace contains a `name` element, which refers to the "name of the namespace".